### PR TITLE
Add metadata to resource file and generated designer implicitly

### PR DESCRIPTION
**Pull request by wli3**
Fix https://github.com/dotnet/sdk/issues/1199

Construct the "link" of the 2 files by using regex since it is
during MSBuild evaluation time. ChangeExtension is not suitable due to
.Designer.cs is not an extension.

Add GetMetaData test helper

### CODE REVIEW

**Comment by dsplaisted** [comment]
> ChangeExtension is not suitable due to .Designer.cs is not an extension.

As suggested in #1441, you can use `Path.GetFileNameWithoutExtension` to go from `foo.Designer.cs` to `foo.Designer`, and then use `Path.ChangeExtension` to change the extension from .Designer to .resx, and you end up with `foo.resx`.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 1]
```diff
@@ -25,7 +25,20 @@ Copyright (c) .NET Foundation. All rights reserved.
   <ItemGroup Condition=" '$(EnableDefaultItems)' == 'true' ">
     <Compile Include="**/*$(DefaultLanguageSourceExtension)" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultCompileItems)' == 'true' " />
     <EmbeddedResource Include="**/*.resx" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultEmbeddedResourceItems)' == 'true' " />
+
+    <EmbeddedResource Update="**\*.resx">
+      <LastGenOutput Condition= "%(LastGenOutput) == ''">$([System.Text.RegularExpressions.Regex]::Replace($([System.IO.Path]::GetFileName(%(Identity))), '\.resx$', '.Designer.cs'))</LastGenOutput>
+      <Generator Condition= "%(Generator) == ''">ResXFileCodeGenerator</Generator>
+    </EmbeddedResource>
+
+    <Compile Update="**\*.Designer.cs">
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: This should also handle VB correctly.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 2]
```diff
@@ -25,7 +25,20 @@ Copyright (c) .NET Foundation. All rights reserved.
   <ItemGroup Condition=" '$(EnableDefaultItems)' == 'true' ">
     <Compile Include="**/*$(DefaultLanguageSourceExtension)" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultCompileItems)' == 'true' " />
     <EmbeddedResource Include="**/*.resx" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultEmbeddedResourceItems)' == 'true' " />
+
+    <EmbeddedResource Update="**\*.resx">
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: need to consider the problem of flags, EnableDefaultEmbeddedResourceItems and EnableDefaultCompileItems

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 3]
```diff
@@ -25,7 +25,20 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+      <LastGenOutput Condition= "%(LastGenOutput) == ''">$([System.Text.RegularExpressions.Regex]::Replace($([System.IO.Path]::GetFileName(%(Identity))), '\.resx$', '.Designer.cs'))</LastGenOutput>
+      <Generator Condition= "%(Generator) == ''">ResXFileCodeGenerator</Generator>
+    </EmbeddedResource>
+
+    <Compile Update="**\*.Designer.cs">
+      <DesignTime Condition= "%(DesignTime) == ''">True</DesignTime>
+      <AutoGen Condition= "%(AutoGen) == ''">True</AutoGen>
+      <DependentUpon Condition= "%(DependentUpon) == ''">$([System.Text.RegularExpressions.Regex]::Replace($([System.IO.Path]::GetFileName(%(Identity))), '\.Designer\.cs$', '.resx'))</DependentUpon>
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: Get file without extension and change extension can replace .Designer.cs

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 4]
```diff
@@ -21,11 +21,28 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+      <_EmbeddedResourceSourceExtension Condition="'$(Language)' == 'VB'">vb</_EmbeddedResourceSourceExtension>
+  </PropertyGroup>
 
   <ItemGroup Condition=" '$(EnableDefaultItems)' == 'true' ">
     <Compile Include="**/*$(DefaultLanguageSourceExtension)" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultCompileItems)' == 'true' " />
     <EmbeddedResource Include="**/*.resx" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultEmbeddedResourceItems)' == 'true' " />
+
+    <EmbeddedResource Update="**\*.resx" Condition="'$(_EmbeddedResourceSourceExtension)' != ''">
+      <LastGenOutput Condition= "%(LastGenOutput) == ''">$([System.IO.Path]::ChangeExtension($([System.IO.Path]::GetFileName(%(Identity))), '.Designer.$(_EmbeddedResourceSourceExtension)'))</LastGenOutput>
+      <Generator Condition= "%(Generator) == ''">ResXFileCodeGenerator</Generator>
```
Comment (sharwell) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: ❓ What does this do for resx files with the code generation option set to None? For example, [this VSPackage.resx](https://github.com/tunnelvisionlabs/LangSvcV2/blob/79d7b636f1dd5bbf175f3a66c6cb31edb3060f5b/Tvl.VisualStudio.Language.Antlr3/Tvl.VisualStudio.Language.Antlr3.csproj#L536-L540) embedded resource should not generate a code-behind file, but the only way to indicate that is by omitting the `Generator` property.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 5]
```diff
@@ -21,11 +21,28 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+    </EmbeddedResource>
+
+    <Compile Update="**\*.Designer.$(_EmbeddedResourceSourceExtension)" Condition="'$(_EmbeddedResourceSourceExtension)' != ''">
```
Comment (sharwell) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: ⚠️ There are many other generated source files which match this pattern, including Windows Forms.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 6]
```diff
@@ -21,11 +21,28 @@ Copyright (c) .NET Foundation. All rights reserved.
     <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
   </PropertyGroup>
 
+  <PropertyGroup>
+      <_EmbeddedResourceSourceExtension Condition="'$(Language)' == 'C#'">cs</_EmbeddedResourceSourceExtension>
```
Comment (sharwell) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: 💭 It seems like this property would already be available by another name.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props [hunk 7]
```diff
@@ -21,11 +21,28 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+      <_EmbeddedResourceSourceExtension Condition="'$(Language)' == 'VB'">vb</_EmbeddedResourceSourceExtension>
+  </PropertyGroup>
 
   <ItemGroup Condition=" '$(EnableDefaultItems)' == 'true' ">
     <Compile Include="**/*$(DefaultLanguageSourceExtension)" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultCompileItems)' == 'true' " />
     <EmbeddedResource Include="**/*.resx" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" Condition=" '$(EnableDefaultEmbeddedResourceItems)' == 'true' " />
+
+    <EmbeddedResource Update="**\*.resx" Condition="'$(_EmbeddedResourceSourceExtension)' != ''">
```
Comment (sharwell) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: ❓ Should this be conditioned on `EnableDefaultEmbeddedResourceItems` as well? Or perhaps some other property?

Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: yes, i haven't address that issue

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props (hunk 5 above)
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: @sharwell good point. And I just talked to Daniel, that sounds like a hard problem. I will set up a meeting for it

**Comment by sharwell** [comment]
@wli3 Here's another implementation you'll want to review for this. It doesn't address all of the above problems I mentioned, but it also points out some others.

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/a70d5e6687ac818a85fe8c9760f0caf5a78dce99/StyleCop.Analyzers/Directory.Build.targets#L4-L17

* This approach is not based on regular expressions, but rather on `Path.ChangeExtension` and `Path.GetFileNameWithoutExtension`
* This approach handles localized resource files

**Comment by wli3** [comment]
Take it down for now, since it is not building

**Pull request closed without merging**